### DISCUSSION TITLE
HttpClient对于x-www-form-urlencoded的支持

### DISCUSSION THREAD

**Discussion by songshitong** [midwayjs/midway#3872] [Q&A] [answered]
### Describe the problem（描述问题）

1. HttpClient直接将data进行json格式化，对于x-www-form-urlencoded不支持
HttpClient对于data的处理逻辑为：
```
async function makeHttpRequest(url, options = {}) {
      if (method === 'GET' && options.data) {
          .....
    }
    else if (options.data) {
        data = Buffer.from(JSON.stringify(options.data));
       .....
    }
}
```

2. 我的请求代码为：
```
 const httpclient = new HttpClient({
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Charset': 'UTF-8',
      },
      timeout: 2000
    });

    let requestResult = await httpclient.request(refreshUrl, { method: 'POST', 
    data:`${encodeURI('script')}=${encodeURI('Jenkins.instance.queue.scheduleMaintenance()')}` });
```
WireShark抓包显示的请求body是Form item: **""script" = "Jenkins.instance.queue.scheduleMaintenance()""** ，同时接口请求失败
![httpclient1](https://github.com/midwayjs/midway/assets/26601878/18e4744f-2e53-4b4a-a590-8f492abc6877)

3. 使用NodeJS的http进行请求
```
const req = request(url.toString(), {
          method:"POST",
          headers:{
             'Content-Type': 'application/x-www-form-urlencoded',
            'Charset': 'UTF-8',
         },
      }, null);

req.end(`${encodeURI('script')}=${encodeURI('Jenkins.instance.queue.scheduleMaintenance()')}`);
```
WireShark的抓包body 是  Form item: **"script" = "Jenkins.instance.queue.scheduleMaintenance()"**，接口可以请求成功

### Midway Versions（Midway 版本）

npx midway-version@latest
Need to install the following packages:
  midway-version@1.3.8

**Comment by czy88840616** [comment, marked as answer]
HttpClient 本意是做一个轻量的 api 发送包，对于 form 这种复杂的流式传输的支持，不在计划内，这种可以使用  node-fetch/axios 来完成。

**Reply by songshitong** [reply]
那接口注释或者文档是否能增加一些限制和说明呢

**Reply by czy88840616** [reply]
![image](https://github.com/midwayjs/midway/assets/418820/bb003073-c108-4bc5-8175-b025a168c9a7)
文档其实也是有描述的
